File: src/main/java/de/steklopod/sec04/Lec02HandleAssignment.java
package de.steklopod.sec04;

import de.steklopod.utils.Util;
import reactor.core.publisher.Flux;

public class Lec02HandleAssignment {

    public static void main(String[] args) {

        Flux.generate(synchronousSink -> synchronousSink.next(Util.faker().country().name()))
                .map(Object::toString)
                .handle((s, synchronousSink) -> {
                    synchronousSink.next(s);
                    if(s.toLowerCase().equals("canada"))
                        synchronousSink.complete();
                })
                .subscribe(Util.subscriber());

    }

}

File: src/main/kotlin/de/steklopod/sec04/Lec02HandleAssignment.kt
package de.steklopod.sec04

import de.steklopod.utils.Util.faker
import de.steklopod.utils.Util.subscriber
import reactor.core.publisher.Flux
import reactor.core.publisher.SynchronousSink
import java.util.*

object Lec02HandleAssignment {
    @JvmStatic
    fun main(args: Array<String>) {
        Flux.generate { synchronousSink: SynchronousSink<Any> -> synchronousSink.next(faker().country().name()) }
            .map { obj: Any -> obj.toString() }
            .handle { s: String, synchronousSink: SynchronousSink<Any> ->
                synchronousSink.next(s)
                if (s.lowercase(Locale.getDefault()) == "canada") synchronousSink.complete()
            }
            .subscribe(subscriber())
    }
}
